### PR TITLE
WT-12708 Avoid null positions from random eviction traversal (#10437) (v5.0 backport)

### PR DESCRIPTION
**Pull request by Siddhartha8899**
To address the problem described in the ticket, the solution is to retry a 100 times until a ref in memory is found.
If not found, then we walk the root page and try to locate the in-mem ref.
This ensures that we mostly avoid returning NULL from the random descent function.

After running wtperf and sys perf, including a high-value workload, I've observed no regression.

---------

Co-authored-by: Siddhartha Mahajan <siddhartha.mahajan@mongodb.com>
(cherry picked from commit 180e5af61e2628799619d5061538ecd96ea6fc7b)

### CODE REVIEW

**Review by ajmorton — COMMENTED** [review]
Looks good on the first pass. 
Can you let me know where the merge conflicts were so I can do an in-depth second pass?

File: src/btree/bt_random.c [hunk 1]
```diff
@@ -350,6 +350,37 @@ __random_leaf(WT_CURSOR_BTREE *cbt)
...
+    cnt = 0;
+    ref_inmem = NULL;
+
+    WT_ASSERT(session, __wt_ref_is_root(current));
+
+    WT_STAT_CONN_INCR(session, cache_eviction_random_sample_inmem_root);
+    WT_STAT_DATA_INCR(session, cache_eviction_random_sample_inmem_root);
+
+    WT_INTL_FOREACH_BEGIN (session, current->page, ref)
+        if (ref->state == WT_REF_MEM) {
```
Comment (Siddhartha8899) [inline src/btree/bt_random.c:373]: There was a merge conflict here
`ref->state` instead of `WT_REF_GET_STATE(ref)`

File: src/btree/bt_random.c [hunk 2]
```diff
@@ -362,14 +393,16 @@ __wt_random_descent(WT_SESSION_IMPL *session, WT_REF **refp, uint32_t flags)
     WT_PAGE *page;
     WT_PAGE_INDEX *pindex;
     WT_REF *current, *descent;
-    uint32_t i, entries, retry;
-    bool eviction;
+    uint32_t i, entries;
+    int retry;
```
Comment (Siddhartha8899) [inline src/btree/bt_random.c:397]: Here, I just had to get rid of new defined variable in later versions - `uint8_t descent_state;` and make retry a signed int.

File: src/btree/bt_random.c [hunk 3]
```diff
@@ -398,6 +431,8 @@ __wt_random_descent(WT_SESSION_IMPL *session, WT_REF **refp, uint32_t flags)
         /* Eviction just wants any random child. */
         if (eviction) {
             descent = pindex->index[__wt_random(&session->rnd) % entries];
+            if (sample_inmem_page && __wt_ref_is_root(current))
+                __random_root_inmem_ref(session, current, &descent, &session->rnd);
```
Comment (Siddhartha8899) [inline src/btree/bt_random.c:435]: Merge conflict - 
In later versions we pass `rnd` to this function. Here we use `&session->rnd`.

**Review by ajmorton — APPROVED** [review]
LGTM. Thanks!!

**Pull request closed without merging**